File: 백준/Silver/2003. 수들의 합 2/수들의 합 2.py
import sys

input = sys.stdin.readline
N, M = map(int, input().split())
nums = list(map(int, input().split()))

left, right = 0, 1
answer = nums.count(M)
while left < right < N:
    sum_of_nums = sum(nums[left:right + 1])
    if sum_of_nums == M:
        answer += 1
        left += 1
        right += 1
    elif sum_of_nums > M:
        right += 1 if left == right - 1 else 0
        left += 1
    else:
        right += 1
print(answer)

File: 백준/Silver/2003. 수들의 합 2/수들의 합 2.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

class Main {
    static int N;
    static long M;
    static int[] arr;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Long.parseLong(st.nextToken());
        arr = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();

        int left = 0, right = 0;
        long sum = arr[0], cnt = 0;
        while (left < N && right < N) {
            if (sum < M) {
                if (right == N - 1) {
                    break;
                }
                sum += arr[++right];
            } else if (sum > M) {
                if (left == right) {
                    if (right == N - 1) {
                        break;
                    }
                    sum += (arr[++right] - arr[left++]);
                } else {
                    sum -= arr[left++];
                }
            } else {
                cnt++;
                if (right == N - 1) {
                    break;
                }
                sum += (arr[++right] - arr[left++]);
            }
        }

        System.out.println(cnt);
    }
}